src/utils/elections/states/pennsylvania.py

```diff
@@ -0,0 +1,188 @@
+"""Download election results from Pennsylvania."""
+
+import json
+import time
+from io import StringIO
+
+import pandas as pd
+import requests
+from utils.constants import BASE_FILEPATH
+
+BASE_URL = "https://www.electionreturns.pa.gov/api/Reports/GenerateReport"
+HEADERS = {
+    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:136.0) Gecko/20100101 Firefox/136.0",
+    "Accept": "application/json, text/plain, */*",
+    "Content-Type": "application/json",
+    "Origin": "https://www.electionreturns.pa.gov",
+    "Referer": "https://www.electionreturns.pa.gov/ReportCenter/Reports",
+}
+TOO_MANY_REQUESTS = 429
+
+
+def fetch_election_data(election_id: int, subtype: str) -> dict:
+    """Fetch election data for a given election ID and subtype.
+
+    Args:
+        election_id: The ID of the election to fetch data for.
+        subtype: The subtype of the election to fetch data for.
+            "G" for General, "P" for Primary, "S" for Special.
+
+    Returns:
+        A pandas DataFrame containing the election data.
+    """
+    payload = {
+        "ElectionID": election_id,
+        "ElectionsubType": subtype,
+        "ExportType": "C",
+        "ReportType": "S",
+        "FileName": "Official",
+    }
+    response = requests.post(BASE_URL, headers=HEADERS, json=payload, timeout=10)
+    if response.status_code == TOO_MANY_REQUESTS:
+        print("Rate limited. Retrying after a delay...")
+        time.sleep(10)
+        return fetch_election_data(election_id, subtype)
+    response.raise_for_status()
+    # Convert the response text to a pandas DataFrame
+    csv_data = StringIO(response.text)
+    election_details = pd.read_csv(csv_data)
+    return election_details
+
+
+def fetch_all_results(elections: list[dict]) -> pd.DataFrame:
+    """Fetch results for all valid elections.
+
+    Args:
+        elections: A list of dictionaries containing election IDs and subtypes.
+
+    Returns:
+        A dataframe containing the election results for all valid elections.
+    """
+    results = []
+    for election in elections:
+        print(
+            f"Fetching results for ElectionID {election['ElectionID']} Subtype {election['Subtype']}..."
+        )
+        try:
+            election_df = fetch_election_data(
+                election["ElectionID"], election["Subtype"]
+            )
+            election_df.loc[:, "election--year"] = election["Year"]
+            results.append(election_df)
+        except Exception as e:
+            print(f"Failed to fetch data for ElectionID {election['ElectionID']}: {e}")
+        time.sleep(1)  # Avoid hitting rate limits
+    return pd.concat(results, ignore_index=True)
+
+
+def get_all_election_ids(start_year: int = None, end_year: int = None) -> list[dict]:
```

Comment:
super nit: since this comes logically before the fetching of election data, I would probably put it at the top.